core/src/main/java/org/elasticsearch/cluster/metadata/PersistedIndexMetaData.java

```diff
@@ -0,0 +1,185 @@
+            parser.nextToken();
+        }
+
+        XContentParser.Token token = parser.currentToken();
+        // first field is not the cluster UUID, so it must be the old way of storing index metadata
+        if (token != XContentParser.Token.FIELD_NAME || parser.currentName().equals(CLUSTER_UUID_FIELD) == false) {
+            // set the cluster UUID with _na_ so we can save the updated index metadata format later
+            return new PersistedIndexMetaData(IndexMetaData.Builder.fromXContent(parser), CLUSTER_UUID_NA_VALUE);
+        }
+
+        if (token != XContentParser.Token.FIELD_NAME) {
+            throw new IllegalArgumentException("expected field name but got a " + parser.currentToken());
+        }
+        String currentFieldName = null;
+        String clusterUUID = null;
+        IndexMetaData indexMetaData = null;
+        do {
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentFieldName = parser.currentName();
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                if (INDEX_METADATA_FIELD.equals(currentFieldName)) {
+                    indexMetaData = IndexMetaData.Builder.fromXContent(parser);
+                } else {
+                    throw new IllegalArgumentException("Unexpected object [" + currentFieldName + "]");
+                }
+            } else if (token.isValue()) {
+                if (CLUSTER_UUID_FIELD.equals(currentFieldName)) {
+                    clusterUUID = parser.text();
+                } else {
+                    throw new IllegalArgumentException("Unexpected field [" + currentFieldName + "]");
+                }
+            } else {
+                throw new IllegalArgumentException("Unexpected token " + token);
+            }
+        } while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT);
+
+        if (indexMetaData == null) {
+            throw new ElasticsearchParseException("Could not parse indexMetaData from x-content");
+        }
+        if (clusterUUID == null) {
+            clusterUUID = PersistedIndexMetaData.CLUSTER_UUID_NA_VALUE;
+        }
+        return new PersistedIndexMetaData(indexMetaData, clusterUUID);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field(CLUSTER_UUID_FIELD, getClusterUUID());
+        builder.startObject(INDEX_METADATA_FIELD);
+        IndexMetaData.Builder.toXContent(getIndexMetaData(), builder, params);
+        builder.endObject();
+        return builder;
+    }
+
+    /**
+     * Test whether the current cluster UUID is the same as the given one.
+     */
+    public boolean isSameClusterUUID(final String otherClusterUUID) {
+        assert clusterUUID != null;
+        assert otherClusterUUID != null;
+        return clusterUUID.equals(otherClusterUUID);
+    }
+
+    @Override
+    public String toString() {
+        try {
```

Comment:
Strings.toString is handy here.. Also, do we really need all the metadata in the tostring? maybe just a summary?